### PR TITLE
feat(release): track prerelease deltas and validate committed notes

### PR DESCRIPTION
**Pull request by ksyasuda**
## Summary

Prerelease notes now include a `Changes since` section for beta/RC deltas, while CI verifies that committed notes match the tagged prerelease version. Release workflows now pass tag values through environment variables, and subtitle fallback behavior, documentation, and tests are updated to reflect the simplified runtime flow.

## Type of change

- [ ] Bug fix
- [x] New feature
- [x] Refactor / internal
- [x] Documentation
- [ ] Other

## How was this tested?

Added focused coverage for prerelease tag selection, fragment delta generation, stale-note detection, marker validation, and real Git repository integration in `scripts/build-changelog.test.ts`. Full local verification has not been run yet.

## Checklist

- [x] Reconciled current-outcome changelog fragment(s), or this PR is labeled `skip-changelog` (see [`changes/README.md`](../changes/README.md))
- [x] Docs updated in the same PR if behavior, defaults, flags, shortcuts, ports, or APIs changed
- [ ] Relevant checks pass locally (typecheck, tests, build)

## Summary by CodeRabbit

- **New Features**
  - Prerelease notes now identify the exact version and summarize changes since the previous prerelease.
  - Changelog generation includes added, modified, and removed change fragments.
  - Added a command to verify prerelease notes match the release version.

- **Bug Fixes**
  - Improved handling of stale or invalid prerelease change sections.
  - Added validation before publishing prerelease releases.

- **Chores**
  - Improved consistency and reliability across release automation.
  - Expanded safeguards for version handling and release publishing.

### CODE REVIEW

**Comment by ksyasuda** [comment]
Addressed the CodeRabbit review, plus the `quality-gate` failure.

**CI failure** (`src/release-workflow.test.ts`): the AUR test pinned the literal `version_no_v="${{ steps.version.outputs.VERSION }}"`, which this branch moved into an `env:` block so the tag reaches the shell as data rather than as script text. Updated the assertion to the new form and kept its original intent (the AppImage name is built from the `v`-stripped version, not the raw tag).

Also added guards for the invariant behind that move, since it is the kind of thing that regresses silently: no tag-derived value may appear inside a `run:` body in `release.yml`, `prerelease.yml`, or `docs-pages.yml`, and the prerelease notes check must run before the release is created or edited.

**Nitpick — resolve fragment paths from the repository root, not `cwd`:** not applying this one. The mechanism is real (git emits repo-root-relative paths), but the predicted outcome — silently omitting fragments and producing incomplete prerelease notes — cannot occur. `writePrereleaseNotesForVersion` calls `readChangeFragments(cwd)` and throws `No changelog fragments found in changes/.` before any delta code runs, so a subdirectory `cwd` fails loudly and early rather than quietly.

I verified this rather than reasoning about it: with a scratch repo, `git diff --name-status <tag> -- changes` run from a subdirectory returns empty (exit 0) because the pathspec resolves against `cwd` — so the proposed fix, which only changes `readCurrent`, would not have addressed the described failure anyway. Writing the subdirectory test made the real behavior explicit: it fails at the fragment read, never reaching the delta.

`cwd` is the project root by contract throughout this script (`package.json`, `changes/`, `CHANGELOG.md`, `release/` all resolve from it), and in this repo that is also the git root. I documented that invariant at `defaultResolveFragmentDelta` instead of adding a `rev-parse` call for an unreachable path.

Verified locally: `bun run test:fast` (577 test files), `bun run typecheck`, `bun run format:check:src`, `bun run changelog:lint`, `bun run test:docs:kb` all pass.

File: src/prerelease-workflow.test.ts [hunk 1]
```diff
@@ -122,3 +122,35 @@ test('prerelease workflow does not publish to AUR', () => {
...
+// tag-derived value used that way is executed as script rather than read as data.
+test('tag-derived values reach shell bodies through env, not template interpolation', () => {
+  const rawVersionUses = prereleaseWorkflow
+    .split('\n')
+    .filter((line) => line.includes('steps.version.outputs.VERSION'))
+    .filter(
+      (line) => !/^\s*RELEASE_VERSION: \$\{\{ steps\.version\.outputs\.VERSION \}\}$/.test(line),
+    );
+  assert.deepEqual(rawVersionUses, []);
+});
```
[the remark this answers is not in the document]
Comment (ksyasuda) [reply src/prerelease-workflow.test.ts:162]: Applied. The finding was correct on all three counts, and I confirmed each evasion against the old guards before replacing them.

The substring allowlist was the core weakness: `${{ steps.version.outputs['VERSION'] }}` does not contain the string `steps.version.outputs.VERSION`, so it never entered the candidate list and the assertion passed on an empty array while the injection was live. Same for `${{ github['ref_name'] }}`. The `indexOf` ordering check was satisfiable by a comment naming the step.

Rather than extend the allowlist to more spellings, the tests now parse the workflows (`Bun.YAML.parse`, via a shared `src/workflow-test-helpers.ts`) and assert on structure:

- `templateExpressionsInRunBodies` rejects **every** `${{ ... }}` in any step `run` body, so alternate spellings are covered by construction rather than enumerated. All three workflows are currently at zero.
- `stepsMissingEnvDeclaration` catches the inverse regression: a step reading `$RELEASE_VERSION` / `$TAG_NAME` whose `env:` block no longer declares it, which would silently expand to empty.
- Ordering compares positions in the parsed step list of the `release` job, so comments and unrelated steps cannot satisfy it. `jobSteps` throws on an unknown job so a rename fails loudly instead of quietly emptying the assertion.

I verified the new guards by reintroducing each of the three evasions from your analysis script into the workflows one at a time; each produced a failing test, and the workflows were restored afterward (`git diff HEAD -- .github/workflows/` is clean).

Verified: `bun run test:fast` (577 test files), `typecheck`, `format:check:src` all pass. Pushed as 615a7857.

File: src/prerelease-workflow.test.ts [hunk 2]
```diff
@@ -122,3 +129,30 @@ test('prerelease workflow does not publish to AUR', () => {
...
+  const publishIndex = steps.findIndex((step) => /gh release (create|edit)/.test(step.run ?? ''));
+
+  assert.notEqual(checkIndex, -1);
+  assert.notEqual(publishIndex, -1);
+  // Stale notes are already published if the check runs after the release.
+  assert.ok(checkIndex < publishIndex);
+  assert.match(
+    steps[checkIndex]!.run!,
+    /changelog:check-prerelease-notes --version "\$RELEASE_VERSION"/,
+  );
```
[the remark this answers is not in the document]
Comment (ksyasuda) [reply src/prerelease-workflow.test.ts]: Confirmed and fixed in 94643ffc.

You were right, and I reproduced it before changing anything: commenting out the real command left the suite fully green (15 pass, 0 fail). My previous fix moved the match from whole-file `indexOf` to a per-step `run` body, but `includes()` still matched inside a shell comment, so the same bypass survived at a narrower scope.

Added `executableRunLines` / `stepRunsCommand` to `src/workflow-test-helpers.ts`, which drop blank and comment-only lines before matching. Both the check step and the publish step are now located that way, and I folded the `--version "$RELEASE_VERSION"` argument into the same match rather than asserting it separately against the raw body, so a commented-out command and a dropped version argument each fail.

Verified both evasions produce a failing test, and the workflows are unmodified afterward (`git diff HEAD -- .github/workflows/` clean). `bun run test:fast` (577 test files), `typecheck`, and `format:check:src` pass.

File: src/prerelease-workflow.test.ts [hunk 3]
```diff
@@ -122,3 +130,28 @@ test('prerelease workflow does not publish to AUR', () => {
...
+    'bun run scripts/build-changelog.ts check-prerelease-notes',
+  );
+
+  // Matched against executable lines only, so commenting the check out fails the
+  // test rather than silently satisfying it.
+  const steps = jobSteps(parsedPrereleaseWorkflow, 'release');
+  const checkIndex = steps.findIndex((step) =>
+    stepRunsCommand(step, /changelog:check-prerelease-notes --version "\$RELEASE_VERSION"/),
+  );
+  const publishIndex = steps.findIndex((step) => stepRunsCommand(step, /gh release (create|edit)/));
```
[the remark this answers is not in the document]
Comment (ksyasuda) [reply src/prerelease-workflow.test.ts]: Confirmed and fixed in 2bfb72a0. Reproduced first: replacing the real check with `echo 'bun run changelog:check-prerelease-notes --version "$RELEASE_VERSION"'` left the suite green at 15 pass, 0 fail.

`stepRunsCommand` now matches at command positions rather than anywhere on a line. `commandPositions` splits each executable line on `;`, `&&`, `||`, and pipes, then strips control-flow prefixes (`if`, `!`, `while`, `then`, ...), so an anchored pattern only matches where a command actually starts. Both workflow patterns are anchored: `/^bun run changelog:check-prerelease-notes --version "\$RELEASE_VERSION"/` and `/^gh release (create|edit)\b/`. Quoted text inside an `echo`/`printf` argument no longer sits at a command position.

Added `src/workflow-test-helpers.test.ts` covering the helper directly, including the `echo` and `printf` regression cases you asked for, plus comment-only lines, a quoted mention inside a redirect, a truncated argument list, and positive cases for `if ! cmd; then`, `set -e && cmd`, and `cmd || exit 1`. The template-expression and env-declaration helpers get unit coverage there too, including bracket spellings and a `$TAGGED` vs `$TAG` word-boundary case.

End-to-end I re-ran all three mention-not-run evasions against the workflow test (`echo`-wrapped, `printf`-wrapped, commented out); each fails, and the workflows are unmodified afterward. `bun run test:fast` (578 test files), `typecheck`, and `format:check:src` pass.

File: src/workflow-test-helpers.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      .map((segment) => {
+        let candidate = segment.trim();
+        let stripped = candidate.replace(COMMAND_PREFIX, '');
+        while (stripped !== candidate) {
+          candidate = stripped;
+          stripped = candidate.replace(COMMAND_PREFIX, '');
+        }
+        return candidate;
+      })
+      .filter(Boolean),
```
[the remark this answers is not in the document]
Comment (ksyasuda) [reply src/workflow-test-helpers.ts:139]: Fixed in 107c6d1b. Reproduced first: `echo 'note; bun run changelog:check-prerelease-notes --version "$RELEASE_VERSION"'` did split inside the quotes and yield a bare `bun run ...` position, so the assertion passed with no validation running. Splitting inside quotes was simply wrong tokenization.

`splitCommandSeparators` now scans character by character tracking quote state, so `;`, `&&`, `||`, and `|` only break a line outside quotes. It stops at an unquoted `#` that opens a word, and it no longer reads the `|` in a redirect such as `2>&1 |` as the start of a new command position.

Regression cases added for quoted `;`, `&&`, and `|` separators, an inline comment after `then`, plus two that guard against over-correcting: a trailing `# comment` must not hide the command in front of it, and a genuine pipe must still yield both positions.

End-to-end, the quoted-separator, echo-wrapped, and commented-out variants each fail the workflow test now, and the workflows are unmodified afterward. `bun run test:fast` (578 test files), `typecheck`, and `format:check:src` pass.

Worth noting where I intend to stop: this helper is a regression guard against ordinary workflow edits, not a sandbox against someone with commit access, who could equally delete the test. It now handles quotes, comments, separators, and redirects correctly. Further hardening (heredocs, subshells, backslash continuations, `eval`) would be chasing full shell parsing for no practical gain, so I plan to leave those.

File: src/workflow-test-helpers.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      current = '';
+      index += 1;
+      continue;
+    }
+    // A lone pipe separates commands; a redirect such as 2>&1 does not.
+    if (char === '|' && !/[0-9<>&]$/.test(current)) {
+      segments.push(current);
+      current = '';
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (ksyasuda) [reply src/workflow-test-helpers.ts:114]: Fixed in db96659b.

Agreed, and the underlying issue was an asymmetry in my own scanner: it already consumed backslash escapes inside double quotes but not outside them, so `echo foo \; bun run verify` split into two positions and reported the second as an executable command.

An unquoted backslash now consumes the following character before the comment and separator checks, which also covers the realistic form of this, `find . -exec ... \;`. Regression cases added for escaped `;` and `|`, the `find -exec` shape, and one guarding the opposite direction: `echo a\\b; bun run verify` must still split on the real separator that follows the escape.

`bun run test:fast` (578 test files), `typecheck`, and `format:check:src` pass.

This is where I am stopping on tokenizer hardening, consistent with the scope you acknowledged on the other thread. The helper now handles quotes, escapes, inline comments, separators, and redirects; heredocs, subshells, line continuations, and `eval` remain out of scope for a workflow regression guard.
